cdap-ui/app/tracker/controllers/audit-ctrl.js

```diff
@@ -118,9 +119,120 @@ class TrackerAuditController {
+      windowTopClass: 'tracker-modal metadata-info-modal',
+      controller: metadataChangeInformation,
+      controllerAs: 'MetadataInfo',
+      resolve: {
+        logDataObj: () => {
+          return logData;
+        }
+      }
+    });
+  }
+}
+
+function metadataChangeInformation(logDataObj, avsc, SchemaHelper) {
+  'ngInject';
+
+  let allFormattedData = {},
+    tags= {},
+    schema= {},
+    properties = {};
+
+  let getSchemaArray = (schemaToParse) => {
+    if(!schemaToParse) {
+      return [];
+    }
+
+    let schemArr = [];
+    let parsed = avsc.parse(schemaToParse, { wrapUnions: true });
+
+    let parsedSchema = parsed.getFields().map((field) => {
+      let type = field.getType();
+      let partialObj = SchemaHelper.parseType(type);
+      return Object.assign({}, partialObj, {
+        id: uuid.v4(),
+        name: field.getName()
+      });
+    });
+
+    parsedSchema.filter((obj) => {
+      if(obj) {
+        schemArr.push(obj.name + ' ' + obj.displayType);
+      }
+    });
+
+    return schemArr;
+  };
+
+  let calculateUpdatedValues = (previous, additions, deletions) => {
+    let result = [];
+    result = _.union(previous, additions);
+    result = _.difference(result, deletions);
+    return result;
+  };
+
+  if(logDataObj.previous.USER) {
+    tags.previousValues = logDataObj.previous.USER.tags;
+    if(logDataObj.additions.USER.tags.length > 0 || logDataObj.deletions.USER.tags.length > 0) {
+      tags.updatedValues = calculateUpdatedValues(logDataObj.previous.USER.tags, logDataObj.additions.USER.tags, logDataObj.deletions.USER.tags);
+    }
+
+    properties.previousValues = _.map(logDataObj.previous.USER.properties, (value, key) =>{
+      return key + ':' + value;
+    });
+
+    properties.updatedValues =_.merge(logDataObj.previous.USER.properties, logDataObj.additions.USER.properties);
+    _.forEach(logDataObj.deletions.USER.properties, function(val, key) {
+      properties.updatedValues = _.omit(properties.updatedValues, key);
+    });
+
+    properties.updatedValues = _.map(properties.updatedValues, (value, key) =>{
+      return key + ':' + value;
+    });
+  }
+
+  if(logDataObj.previous.SYSTEM) {
+    tags.previousValues = logDataObj.previous.SYSTEM.tags;
+    if(logDataObj.additions.SYSTEM.tags.length > 0 || logDataObj.deletions.SYSTEM.tags.length > 0) {
+      tags.updatedValues = calculateUpdatedValues(logDataObj.previous.SYSTEM.tags, logDataObj.additions.SYSTEM.tags, logDataObj.deletions.SYSTEM.tags);
+    }
+
+    if(logDataObj.previous.SYSTEM.properties.schema) {
+      schema.previousValues = getSchemaArray(logDataObj.previous.SYSTEM.properties.schema);
+    }
+
+    let schemaArrayAddition = getSchemaArray(logDataObj.additions.SYSTEM.properties.schema);
+    let schemaArrayDeletion = getSchemaArray(logDataObj.deletions.SYSTEM.properties.schema);
+
+    schema.updatedValues = calculateUpdatedValues(schema.previousValues, schemaArrayAddition, schemaArrayDeletion);
+  }
```

Comment:
this block of code for `SYSTEM` seems to be very similar to `USER`. Can this be refactored into a helper function and consolidate the code for `SYSTEM` and `USER`?